### PR TITLE
Merge CUDA integration scheme into development

### PR DESCRIPTION
**Pull request by dwillcox**
This merges the work done on the cudadevice branch into development. The cudadevice-copy branch is the cudadevice branch rebased on top of development.

This works with FBoxLib and cuda 9 on summitdev, but not cuda 8.

Most of the intrusive refactoring is in the VODE90 integrator, with minor modification to the other components of Microphysics to mark various subroutines as `attributes(device)`.

### CODE REVIEW

File: rates/aprox_rates.F90 [hunk 1]
```diff
@@ -24,45 +28,34 @@ subroutine rates_init()
...
-        		      0.6673, 0.6964, 1.0985, 2.2849, 4.0614, 6.1033, &
-        		      0.9009, 0.9175, 1.2525, 2.3619, 4.0882, 6.1099, &
-        		      1.1032, 1.113, 1.3947, 2.4345, 4.1161, 6.1171 /), &
-                           (/ 6, 14 /) )
-
-
+    tv = (/ 2.0, 4.0, 6.0, 8.0, 10.0, 12.0, 14.0 /)
+
+    datn(1,:,:) = reshape( (/ -3.98, -2.84, -1.41,  0.20,  1.89,  3.63, &
+                              -3.45, -2.62, -1.32,  0.22,  1.89,  3.63, &
```
Comment (zingale) [inline rates/aprox_rates.F90]: why are these different?

File: rates/aprox_rates.F90 [hunk 2]
```diff
@@ -71,7 +64,7 @@ subroutine rates_init()
        rfd2(k)=1./((rv(k+2)-rv(k-1))*(rv(k+2)-rv(k))*(rv(k+2)-rv(k+1)))
     enddo
 
-    do j = 2, 12
+    do j = 2, 5
```
Comment (zingale) [inline rates/aprox_rates.F90]: why did this change?

File: rates/aprox_rates.F90 [hunk 3]
```diff
@@ -154,119 +149,10 @@ subroutine rate_c12ag(tf,den,fr,dfrdt,dfrdd,rr,drrdt,drrdd)
 
   end subroutine rate_c12ag
 
-  ! This routine computes the nuclear reaction rate for 12C(a,g)16O and its inverse
```
Comment (zingale) [inline rates/aprox_rates.F90]: why is this gone?

File: rates/aprox_rates.F90 [hunk 4]
```diff
@@ -2695,17 +2681,17 @@ subroutine langanke(btemp,bden,y56,ye,rn56ec,sn56ec)
...
-    double precision :: rnt(2),rne(2,14),t9,r,rfm,rf0, &
+    double precision :: rnt(2),rne(2,7),t9,r,rfm,rf0, &
                         rf1,rf2,dfacm,dfac0,dfac1,dfac2, &
                         tfm,tf0,tf1,tf2,tfacm,tfac0,tfac1,tfac2
 
     ! calculate ni56 electron capture and neutrino loss rates
     rn56ec = 0.0
     sn56ec = 0.0
-    if ( (btemp .lt. 1.0e9) .or. (bden*ye .lt. 1.0e6)) return
+    if ( (btemp .lt. 2.0e9) .or. (bden*ye .lt. 1.0e6)) return
```
Comment (zingale) [inline rates/aprox_rates.F90]: why is this different?

File: rates/aprox_rates.F90 [hunk 5]
```diff
@@ -2695,17 +2681,17 @@ subroutine langanke(btemp,bden,y56,ye,rn56ec,sn56ec)
...
     t9    = min(btemp,1.4d10) * 1.0d-9
     r     = max(6.0d0,min(11.0d0,log10(bden*ye)))
-    jp    = min(max(2,int(t9)),12)
+    jp    = min(max(2,int(0.5d0*t9)),5)
```
Comment (zingale) [inline rates/aprox_rates.F90]: different?

File: rates/aprox_rates.F90 [hunk 6]
```diff
@@ -2736,11 +2722,9 @@ subroutine langanke(btemp,bden,y56,ye,rn56ec,sn56ec)
 
     ! set the output
     rn56ec = 10.0d0**rnt(1)
-    sn56ec = 6.022548d+23 * 1.60218d-6 * y56 * 10.0d0**rnt(2)
-   
-    !write(*,*) "btemp",btemp, "bden", bden, "t9",t9,"r",r, "rn56ec",rn56ec, "sn56ec", sn56ec
+    sn56ec = 6.022548d+23 * 8.18683d-7 * y56 * 10.0d0**rnt(2)
```
Comment (zingale) [inline rates/aprox_rates.F90]: why changed?

File: unit_test/burn_cell/main.f90 [hunk 7]
```diff
@@ -78,7 +78,7 @@ program burn_cell
   burn_state_in%xn(:) = massfractions(:)
 
   ! normalize -- just in case
-  !call normalize_abundances_burn(burn_state_in)
+  call normalize_abundances_burn(burn_state_in)
```
Comment (zingale) [inline unit_test/burn_cell/main.f90]: why did this change?

File: unit_test/test_react/inputs_aprox13 [hunk 8]
```diff
@@ -10,7 +10,9 @@
 
   tmax = 1.d-3
 
-  xin_file   = "xin.aprox13"
-  run_prefix = "react_aprox13_"
+  call_eos_in_rhs = T
```
Comment (zingale) [inline unit_test/test_react/inputs_aprox13]: this will change answers

File: unit_test/test_react/inputs_ignition.16 [hunk 9]
```diff
@@ -8,6 +8,8 @@
 
   tmax = 1.0d-1
 
-  xin_file   = "xin.ignition"
-  run_prefix = "react_ignition_"
+  call_eos_in_rhs = T
```
Comment (zingale) [inline unit_test/test_react/inputs_ignition.16]: this will change answers

File: unit_test/test_react/inputs_reaclib_cburn.16 [hunk 10]
```diff
@@ -21,6 +21,8 @@
   retry_burn = .false.
   renormalize_abundances = .false.
 
+  call_eos_in_rhs = T
```
Comment (zingale) [inline unit_test/test_react/inputs_reaclib_cburn.16]: this will change answers

File: unit_test/test_react/main.F90 [hunk 11]
```diff
@@ -159,101 +183,72 @@ program test_react
...
-           enddo
-        enddo
-     enddo
-     !$acc end parallel
-     !$acc end data
-
-     !$OMP END PARALLEL DO
-
-     ! End the timer and print the results.
+     write(*,*) 'lo = ', lo
```
Comment (zingale) [inline unit_test/test_react/main.F90]: we still want an OpenMP version, right?

File: integration/VODE90/vode_rhs.F90 [hunk 12]
```diff
@@ -38,9 +44,14 @@ subroutine f_rhs(neq, time, y, ydot, rpar, ipar)
     ydot = ZERO
 
     ! Fix the state as necessary.
-
     call clean_state(y, rpar)
 
+
+    ! Renormalize the abundances as necessary.
+    if (renormalize_abundances) then
```
[the remark this answers is not in the document]
Comment (dwillcox) [reply integration/VODE90/vode_rhs.F90]: In the both branches this call is already in clean_state. I've removed this duplicate call.

File: integration/VODE90/actual_integrator.F90 [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+    else if (burning_mode == 1 .or. burning_mode == 3) then
+       dvode_state % rpar(irp_self_heat) = ONE
+    else
+       stop
+       !CUDA
+       !call bl_error("Error: unknown burning_mode in actual_integrator.f90.")
+    endif
+
+
+    if (burning_mode == 0 .or. burning_mode == 2) then
```
[the remark this answers is not in the document]
Comment (dwillcox) [reply integration/VODE90/actual_integrator.F90]: indeed, I've removed one of the duplicates.

File: interfaces/eos_type.F90 [hunk 14]
```diff
@@ -39,22 +39,27 @@ module eos_type_module
...
-  real(rt), save :: maxye   = 1.d0 + 1.d-12
-  real(rt), save :: mine    = 1.d-200
-  real(rt), save :: maxe    = 1.d200
-  real(rt), save :: minp    = 1.d-200
-  real(rt), save :: maxp    = 1.d200
-  real(rt), save :: mins    = 1.d-200
-  real(rt), save :: maxs    = 1.d200
-  real(rt), save :: minh    = 1.d-200
-  real(rt), save :: maxh    = 1.d200
+  real(rt), allocatable, save :: mintemp
```
[the remark this answers is not in the document]
Comment (dwillcox) [reply interfaces/eos_type.F90]: we set those in unit_test/eos.F90 in the eos_init subroutine after allocating them.

File: unit_test/burn_cell/main.f90 (hunk 7 above)
Comment (dwillcox) [reply unit_test/burn_cell/main.f90]: I probably commented it out for testing .... not sure why offhand.

For generality, I've added `normalize_input_species` as a probin variable defaulting to .false. to maintain behavior on the development branch.

File: unit_test/test_react/inputs_aprox13 (hunk 8 above)
Comment (dwillcox) [reply unit_test/test_react/inputs_aprox13]: Yes, I switched to calling the eos in the rhs for all my GPU tests.

File: unit_test/test_react/inputs_ignition.16 (hunk 9 above)
Comment (dwillcox) [reply unit_test/test_react/inputs_ignition.16]: Yes, I switched to calling the eos in the rhs for all my GPU tests.

File: unit_test/test_react/inputs_reaclib_cburn.16 (hunk 10 above)
Comment (dwillcox) [reply unit_test/test_react/inputs_reaclib_cburn.16]: Yes, I switched to calling the eos in the rhs for all my GPU tests.

File: unit_test/test_react/main.F90 (hunk 11 above)
Comment (dwillcox) [reply unit_test/test_react/main.F90]: I moved the OpenMP into the react_zones subroutine in react_zones_module.F90 which does the actual looping over zones.

File: rates/aprox_rates.F90 (hunk 1 above)
Comment (dwillcox) [reply rates/aprox_rates.F90]: Carl changed these in 5b392cd9 in development, I think the rebase must have messed this up.

I'll go with the version in development.

**Comment by zingale** [comment]
this is being broken up

**Pull request closed without merging**